File: src/utils/serviceRadius.ts

import { CellData, BuildingType, ConstructionStatus, ServiceCoverage, GRID_SIZE } from '../types';
import { BUILDINGS, VARIANTS } from '../data/gameData';

/**
 * 计算所有建筑的服务覆盖范围
 * 使用曼哈顿距离，以建筑中心为圆心辐射
 */
export const calculateServiceCoverage = (grid: CellData[][]): void => {
  // 清除所有服务覆盖标记
  for (let y = 0; y < GRID_SIZE; y++) {
    for (let x = 0; x < GRID_SIZE; x++) {
      grid[y][x].serviceCoverage = undefined;
    }
  }

  // 遍历所有已完工且有服务半径的建筑
  for (let y = 0; y < GRID_SIZE; y++) {
    for (let x = 0; x < GRID_SIZE; x++) {
      const cell = grid[y][x];
      if (!cell.isOrigin) continue;
      if (cell.constructionStatus !== ConstructionStatus.COMPLETED) continue;

      const def = BUILDINGS[cell.building];
      if (!def.serviceType || !def.serviceRadius) continue;

      // 获取建筑尺寸以计算中心点
      const variants = VARIANTS[cell.building];
      const variant = variants?.find(v => v.id === cell.variantId);
      const bw = variant ? variant.width : (def.width || 1);
      const bh = variant ? variant.height : (def.height || 1);
      const w = cell.rotation ? bh : bw;
      const h = cell.rotation ? bw : bh;

      const centerX = x + w / 2;
      const centerY = y + h / 2;
      const radius = def.serviceRadius;
      const serviceType = def.serviceType;

      // 曼哈顿距离覆盖
      const minY = Math.max(0, Math.floor(centerY - radius));
      const maxY = Math.min(GRID_SIZE - 1, Math.ceil(centerY + radius));
      const minX = Math.max(0, Math.floor(centerX - radius));
      const maxX = Math.min(GRID_SIZE - 1, Math.ceil(centerX + radius));

      for (let ty = minY; ty <= maxY; ty++) {
        for (let tx = minX; tx <= maxX; tx++) {
          const dist = Math.abs(tx + 0.5 - centerX) + Math.abs(ty + 0.5 - centerY);
          if (dist <= radius) {
            const target = grid[ty][tx];
            if (!target.serviceCoverage) {
              target.serviceCoverage = {};
            }
            target.serviceCoverage[serviceType] = true;
          }
        }
      }
    }
  }
};

/**
 * 获取某个建筑的服务覆盖统计
 */
export const getBuildingServiceInfo = (
  grid: CellData[][], originX: number, originY: number
): { food: boolean; study: boolean; recreation: boolean } => {
  const cell = grid[originY][originX];
  const def = BUILDINGS[cell.building];
  const variants = VARIANTS[cell.building];
  const variant = variants?.find(v => v.id === cell.variantId);
  const bw = variant ? variant.width : (def.width || 1);
  const bh = variant ? variant.height : (def.height || 1);
  const w = cell.rotation ? bh : bw;
  const h = cell.rotation ? bw : bh;

  let food = false, study = false, recreation = false;

  // 检查建筑占地范围内是否有服务覆盖
  for (let dy = 0; dy < h; dy++) {
    for (let dx = 0; dx < w; dx++) {
      const t = grid[originY + dy]?.[originX + dx];
      if (t?.serviceCoverage) {
        if (t.serviceCoverage.food) food = true;
        if (t.serviceCoverage.study) study = true;
        if (t.serviceCoverage.recreation) recreation = true;
      }
    }
  }

  return { food, study, recreation };
};

File: MyUniversity.swiftpm/Sources/ServiceRadius.swift
import Foundation

enum ServiceRadius {
    static func calculateServiceCoverage(_ grid: inout [[CellData]]) {
        // Clear
        for y in 0..<GRID_SIZE {
            for x in 0..<GRID_SIZE {
                grid[y][x].serviceCoverage = nil
            }
        }

        // Find service buildings
        for y in 0..<GRID_SIZE {
            for x in 0..<GRID_SIZE {
                let cell = grid[y][x]
                guard cell.isOrigin else { continue }
                guard cell.constructionStatus == .completed else { continue }

                guard let def = BUILDINGS[cell.building] else { continue }
                guard let serviceType = def.serviceType, let serviceRadius = def.serviceRadius else { continue }

                let variants = VARIANTS[cell.building]
                let variant = variants?.first(where: { $0.id == cell.variantId })
                let bw = variant?.width ?? def.width
                let bh = variant?.height ?? def.height
                let w = cell.rotation != 0 ? bh : bw
                let h = cell.rotation != 0 ? bw : bh

                let centerX = Double(x) + Double(w) / 2.0
                let centerY = Double(y) + Double(h) / 2.0
                let radius = Double(serviceRadius)

                let minY = max(0, Int(floor(centerY - radius)))
                let maxY = min(GRID_SIZE - 1, Int(ceil(centerY + radius)))
                let minX = max(0, Int(floor(centerX - radius)))
                let maxX = min(GRID_SIZE - 1, Int(ceil(centerX + radius)))

                for ty in minY...maxY {
                    for tx in minX...maxX {
                        let dist = abs(Double(tx) + 0.5 - centerX) + abs(Double(ty) + 0.5 - centerY)
                        if dist <= radius {
                            if grid[ty][tx].serviceCoverage == nil {
                                grid[ty][tx].serviceCoverage = ServiceCoverage()
                            }
                            switch serviceType {
                            case .food: grid[ty][tx].serviceCoverage?.food = true
                            case .study: grid[ty][tx].serviceCoverage?.study = true
                            case .recreation: grid[ty][tx].serviceCoverage?.recreation = true
                            }
                        }
                    }
                }
            }
        }
    }

    static func getBuildingServiceInfo(_ grid: [[CellData]], originX: Int, originY: Int) -> ServiceCoverage {
        let cell = grid[originY][originX]
        guard let def = BUILDINGS[cell.building] else { return ServiceCoverage() }
        let variants = VARIANTS[cell.building]
        let variant = variants?.first(where: { $0.id == cell.variantId })
        let bw = variant?.width ?? def.width
        let bh = variant?.height ?? def.height
        let w = cell.rotation != 0 ? bh : bw
        let h = cell.rotation != 0 ? bw : bh

        var result = ServiceCoverage()
        for dy in 0..<h {
            for dx in 0..<w {
                let ny = originY + dy
                let nx = originX + dx
                guard ny < GRID_SIZE, nx < GRID_SIZE else { continue }
                if let cov = grid[ny][nx].serviceCoverage {
                    if cov.food { result.food = true }
                    if cov.study { result.study = true }
                    if cov.recreation { result.recreation = true }
                }
            }
        }
        return result
    }
}
